sway-lib-std/src/storage/storage_vec.sw

```diff
@@ -939,6 +940,61 @@ impl<V> StorageKey<StorageVec<V>> {
             }
         }
     }
+
+    /// Returns a `StorageVecIter` to iterate over this `StorageVec`.
+    ///
+    /// # Returns
+    ///
+    /// * [StorageVecIter<V>] - The struct which can be iterated over.
+    ///
+    /// # Examples
+    ///
+    /// ```sway
+    /// storage {
+    ///     vec: StorageVec<u64> = StorageVec {},
+    /// }
+    ///
+    /// fn foo() {
+    ///     storage.vec.push(5);
+    ///     storage.vec.push(10);
+    ///     storage.vec..push(15);
+    ///
+    ///     // Get the iterator
+    ///     let iter = storage.vec.iter();
+    ///
+    ///     assert(5 == iter.next().unwrap().read());
+    ///     assert(10 == iter.next().unwrap().read());
+    ///     assert(15 == iter.next().unwrap().read());
+    /// }
+    /// ```
+    pub fn iter(self) -> StorageVecIter<V> {
+        StorageVecIter {
+            values: self,
+            index: 0,
+        }
+    }
+}
+
+pub struct StorageVecIter<V> {
+    values: StorageKey<StorageVec<V>>,
+    index: u64,
+}
+
+impl<V> StorageIterator for StorageVecIter<V> {
+    type Item = StorageKey<V>;
+    #[storage(read)]
+    fn next(ref mut self) -> Option<Self::Item> {
+        if self.index >= self.values.len() {
+            return None
+        }
+
+        self.index += 1;
+        let key = self.values.get(self.index - 1);
```

Comment:
Also, speaking of the performance, if we provide iterator over `StorageVec` I would love it to be not only a convenience, but also the most performant way to iterate the vector.

E.g., we can consider an optimization which minimizes the storage reads by keeping already read elements. This will happen in case of storing types whose size is less then 32 bytes. E.g., we can fetch four `u64`s with one read and keep them internally in the `StorageVecIter` to minimize further reads.

This will of course make the implementation of `next` more complex which also costs gas. The question is in the difference of the gas spent on storage access and internal implementation of `storage_api::read` (which is already complex) compared to more complex implementation that minimizes reads, allocs, etc.